agent/src/main/java/org/jfrog/teamcity/agent/util/BuildInfoUtils.java

```diff
@@ -0,0 +1,185 @@
+import jetbrains.buildServer.agent.Constants;
+import jetbrains.buildServer.agent.artifacts.ArtifactsWatcher;
+import jetbrains.buildServer.util.ArchiveUtil;
+import org.apache.commons.lang.StringUtils;
+import org.jfrog.build.api.Agent;
+import org.jfrog.build.api.Build;
+import org.jfrog.build.api.BuildInfoProperties;
+import org.jfrog.build.api.BuildRetention;
+import org.jfrog.build.api.builder.BuildInfoBuilder;
+import org.jfrog.build.client.ProxyConfiguration;
+import org.jfrog.build.extractor.BuildInfoExtractorUtils;
+import org.jfrog.build.extractor.clientConfiguration.ArtifactoryBuildInfoClientBuilder;
+import org.jfrog.build.extractor.clientConfiguration.ArtifactoryDependenciesClientBuilder;
+import org.jfrog.build.extractor.clientConfiguration.IncludeExcludePatterns;
+import org.jfrog.build.extractor.clientConfiguration.PatternMatcher;
+import org.jfrog.build.extractor.clientConfiguration.client.ArtifactoryBuildInfoClient;
+import org.jfrog.teamcity.agent.ServerConfig;
+import org.jfrog.teamcity.common.RunnerParameterKeys;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.jfrog.teamcity.common.ConstantValues.*;
+
+/**
+ * Created by Bar Belity on 26/10/2020.
+ */
+public class BuildInfoUtils {
+
+    /**
+     * Create gz of the build info and publish it to the server. The file will be saved under
+     * .BuildServer/system/artifacts/$Project/$Build/$BuildNumber/.teamcity/artifactory-build-info.json.gz
+     */
+    public static void publishBuildInfoToTeamCityServer(AgentRunningBuild build, Build buildInfo, ArtifactsWatcher watcher) {
+        try {
+            File buildInfoFile = new File(build.getAgentTempDirectory(), BUILD_INFO_FILE_NAME);
+            BuildInfoExtractorUtils.saveBuildInfoToFile(buildInfo, buildInfoFile);
+            File buildInfoPacked = ArchiveUtil.packFile(buildInfoFile);
+            watcher.addNewArtifactsPath(buildInfoPacked.getAbsolutePath() + "=>.teamcity");
+
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to publish build info on TeamCity server", e);
+        }
+    }
+
+    public static void sendBuildAndBuildRetention(AgentRunningBuild build, Build buildInfo, BuildRetention retention, boolean isAsyncRetention, ArtifactoryBuildInfoClient infoClient)
+            throws Exception {
+        try {
+            build.getBuildLogger().progressMessage("Deploying build info ...");
+            org.jfrog.build.extractor.retention.Utils.sendBuildAndBuildRetention(infoClient, buildInfo, retention, isAsyncRetention);
+            build.getBuildLogger().progressFinished();
+        } catch (Exception e) {
+            throw new Exception("Error deploying Artifactory build-info.", e);
+        }
+    }
+
+    public static BuildInfoBuilder getBuildInfoBuilder(Map<String, String> runnerParams, BuildRunnerContext runnerContext) {
```

Comment:
We should remove the old ```getBuildInfoBuilder``` method, located in the ```BaseBuildInfoExtractor``` class.

Reply:
The method `getBuildInfoBuilder` also adds `LicenseControl` and `Governance`, which are deprecated.
Not using the existing `getBuildInfoBuilder` requires a wide change of removing these deprecated configuration also from the tasks UI etc.
In my opinion it would be better if we do it in a separate PR.
What do you think?

Reply:
After discussion we decided to remove the deprecated behaviour in a separated PR.